File: Algorithms/Easy/UniqueEmailAddresses.go
package easy

// Time complexity: O(m * n)
// Space complexity: O(m)
func numUniqueEmails(emails []string) int {
	seen := make(map[string]struct{})
	for _, email := range emails {
		idx := atSignIdx(email)
		validEmail := make([]byte, 0, len(email))
		for i := 0; i < len(email); i++ {
			if email[i] == '+' || email[i] == '@' {
				break
			}
			if email[i] == '.' {
				continue
			}
			validEmail = append(validEmail, email[i])
		}

		validEmail = append(validEmail, email[idx:]...)
		seen[string(validEmail)] = struct{}{}
	}

	return len(seen)
}

func atSignIdx(email string) int {
	for i := 0; i < len(email); i++ {
		if email[i] == '@' {
			return i
		}
	}

	return -1
}

File: Algorithms/Easy/929.Unique-Email-Addresses/UniqueEmailAddresses.py
# Time complexity: O(m*n)
# Space complexity: O(m)
class Solution:
    def numUniqueEmails(self, emails: list[str]) -> int:
        unique_email = set()
        for email in emails:
            splitted = email.split("@")
            splitted[0] = splitted[0].replace(".", "")
            idx = splitted[0].find("+")
            if idx != -1:
                splitted[0] = splitted[0][:idx]

            unique_email.add("@".join(splitted))

        return len(unique_email)
